File: fixilya-backend/fixilya-microservices/services/user-service/src/services/user.service.js
const { AppError } = require("../../../../shared/utils/appError");
const admin = require("../../../../shared/config/firebase");
const logger = require("../../../../shared/utils/logger");
const {
  HANDYMAN_COLLECTION_NAME,
  CLIENT_COLLECTION_NAME,
  USER_COLLECTION_NAME,
  ADMIN_NOTIFICATIONS_COLLECTION_NAME,
} = require("../../../../shared/config/appConstats");

const auth = admin.auth();
const db = admin.firestore();

class UserService {
  // ─────────────────────────────────────────────
  // PROFILE COMPLETION (called right after registration — no auth token yet)
  // ─────────────────────────────────────────────

  /**
   * Creates a minimal profile document so the user can proceed without
   * filling out their full profile immediately.
   * Handymen will need to complete their profile later to get approved.
   */
  async completeProfileWithSkip({ uid, userType }) {
    try {
      logger.info(`📝 [completeProfileWithSkip] uid=${uid} userType=${userType}`);

      const userRecord = await auth.getUser(uid);
      if (!userRecord) {
        throw new AppError("User not found", 404);
      }

      const normalizedUserType = userType === "customer" ? "client" : userType;
      const collection =
        normalizedUserType === "handyman"
          ? HANDYMAN_COLLECTION_NAME
          : CLIENT_COLLECTION_NAME;

      // Track whether this is the first time the doc is created
      const collectionDoc = await db.collection(collection).doc(uid).get();
      const isNewUser = !collectionDoc.exists;

      const timestamp = new Date().toISOString();

      const profileData =
        normalizedUserType === "handyman"
          ? {
              uid,
              email: userRecord.email || "",
              fullName: userRecord.displayName || "",
              phone: userRecord.phoneNumber || "",
              city: "",
              experience: "0",
              hourlyRate: 0,
              bio: "",
              skills: [],
              profilePicture: userRecord.photoURL || "",
              workImages: [],
              profileCompleted: false,
              approved: false,
              rejected: false,
              suspended: false,
              reviewStatus: "pending",
              rating: 0,
              totalReviews: 0,
              completedJobs: 0,
              createdAt: timestamp,
              updatedAt: timestamp,
            }
          : {
              uid,
              email: userRecord.email || "",
              fullName: userRecord.displayName || "",
              phone: userRecord.phoneNumber || "",
              city: "",
              profilePicture: userRecord.photoURL || "",
              profileCompleted: false,
              approved: true,
              suspended: false,
              createdAt: timestamp,
              updatedAt: timestamp,
            };

      // merge: true keeps any fields already saved (e.g. from a previous attempt)
      await db.collection(collection).doc(uid).set(profileData, { merge: true });
      logger.info(`✅ Profile created in ${collection} for uid=${uid}`);

      await db.collection(USER_COLLECTION_NAME).doc(uid).update({
        profileCompleted: false,
        updatedAt: timestamp,
      });

      await auth.setCustomUserClaims(uid, {
        userType: normalizedUserType,
        role: normalizedUserType,
      });
      logger.info(`✅ Custom claims set for uid=${uid}`);

      // Notify admin only for brand-new handymen
      if (normalizedUserType === "handyman" && isNewUser) {
        await this._notifyAdminNewHandyman({
          handymanId: uid,
          handymanName: userRecord.displayName || "Unknown",
          email: userRecord.email || "",
          phone: userRecord.phoneNumber || "",
          profileCompleted: false,
          timestamp,
        });
      }

      return {
        success: true,
        message:
          normalizedUserType === "handyman"
            ? "Profile created. Complete it later to get approved."
            : "Profile created successfully.",
        data: {
          uid,
          userType: normalizedUserType,
          profileCompleted: false,
          needsApproval: normalizedUserType === "handyman",
        },
      };
    } catch (error) {
      logger.error(`❌ [completeProfileWithSkip] ${error.message}`);
      throw error;
    }
  }

  /**
   * Saves a full profile (all fields) submitted from the onboarding screen.
   * Accepts Cloudinary URLs for profilePicture and workImages.
   */
  async completeProfile({
    uid,
    userType,
    fullName,
    phone,
    city,
    experience,
    bio,
    skills,
    profilePicture,
    workImages,
  }) {
    try {
      logger.info(`📝 [completeProfile] uid=${uid} userType=${userType}`);

      const firebaseUser = await auth.getUser(uid);
      if (!firebaseUser) {
        throw new AppError("User not found", 404);
      }

      const normalizedUserType = userType === "customer" ? "client" : userType;
      const collection =
        normalizedUserType === "handyman"
          ? HANDYMAN_COLLECTION_NAME
          : CLIENT_COLLECTION_NAME;

      const isNewUser = await this._isFirstTimeUser(uid);
      const timestamp = new Date().toISOString();

      const profileData =
        normalizedUserType === "handyman"
          ? {
              uid,
              email: firebaseUser.email || "",
              fullName: fullName || firebaseUser.displayName || "",
              phone: phone || firebaseUser.phoneNumber || "",
              city: city || "",
              experience: experience || "0",
              bio: bio || "",
              skills: skills || [],
              profilePicture: profilePicture || "",
              workImages: workImages || [],
              profileCompleted: true,
              approved: false,
              rejected: false,
              suspended: false,
              reviewStatus: "pending",
              rating: 0,
              totalReviews: 0,
              completedJobs: 0,
              createdAt: timestamp,
              updatedAt: timestamp,
            }
          : {
              uid,
              email: firebaseUser.email || "",
              fullName: fullName || firebaseUser.displayName || "",
              phone: phone || firebaseUser.phoneNumber || "",
              city: city || "",
              profilePicture: profilePicture || "",
              profileCompleted: true,
              approved: true,
              suspended: false,
              createdAt: timestamp,
              updatedAt: timestamp,
            };

      await db.collection(collection).doc(uid).set(profileData, { merge: true });
      logger.info(`✅ Full profile saved to ${collection} for uid=${uid}`);

      await db.collection(USER_COLLECTION_NAME).doc(uid).update({
        profileCompleted: true,
        fullName: fullName || firebaseUser.displayName || "",
        phone: phone || firebaseUser.phoneNumber || "",
        updatedAt: timestamp,
      });

      await auth.setCustomUserClaims(uid, {
        userType: normalizedUserType,
        role: normalizedUserType,
      });

      if (normalizedUserType === "handyman" && isNewUser) {
        await this._notifyAdminNewHandyman({
          handymanId: uid,
          handymanName: fullName || firebaseUser.displayName || "Unknown",
          email: firebaseUser.email || "",
          phone: phone || firebaseUser.phoneNumber || "",
          profileCompleted: true,
          timestamp,
        });
      }

      logger.info(
        `✅ Full profile complete for uid=${uid} | pic=${profilePicture ? "yes" : "no"} | workImages=${workImages?.length ?? 0}`
      );

      return {
        success: true,
        message:
          normalizedUserType === "handyman"
            ? isNewUser
              ? "Your profile is under review. You'll be notified once approved."
              : "Your profile has been updated successfully."
            : "Profile created successfully.",
        data: {
          uid,
          userType: normalizedUserType,
          profileCompleted: true,
          needsApproval: normalizedUserType === "handyman",
          isNewUser,
          profilePictureUrl: profilePicture || "",
          workImageUrls: workImages || [],
        },
      };
    } catch (error) {
      logger.error(`❌ [completeProfile] ${error.message}`);
      throw error;
    }
  }

  // ─────────────────────────────────────────────
  // GENERAL PROFILE (any user type)
  // ─────────────────────────────────────────────

  /**
   * Returns merged profile from users + type-specific collection.
   */
  async getProfile(uid) {
    try {
      const userDoc = await db.collection(USER_COLLECTION_NAME).doc(uid).get();

      if (!userDoc.exists) {
        throw new AppError("User not found", 404);
      }

      const userData = userDoc.data();
      const userType = userData.userType;
      const collection =
        userType === "handyman" ? HANDYMAN_COLLECTION_NAME : CLIENT_COLLECTION_NAME;

      const profileDoc = await db.collection(collection).doc(uid).get();

      return {
        ...userData,
        ...(profileDoc.exists ? profileDoc.data() : {}),
      };
    } catch (error) {
      logger.error(`❌ [getProfile] uid=${uid}: ${error.message}`);
      throw error;
    }
  }

  /**
   * Updates general profile fields. Mirrors name / phone into the type-specific collection.
   */
  async updateProfile(uid, updates) {
    try {
      const userDoc = await db.collection(USER_COLLECTION_NAME).doc(uid).get();

      if (!userDoc.exists) {
        throw new AppError("User not found", 404);
      }

      const userData = userDoc.data();
      const collection =
        userData.userType === "handyman"
          ? HANDYMAN_COLLECTION_NAME
          : CLIENT_COLLECTION_NAME;

      // Strip immutable fields
      const protectedFields = ["uid", "email", "userType", "createdAt"];
      protectedFields.forEach((field) => delete updates[field]);

      updates.updatedAt = new Date().toISOString();

      await db.collection(collection).doc(uid).update(updates);

      // Mirror name / phone into users collection
      const userUpdates = {};
      if (updates.fullName) userUpdates.fullName = updates.fullName;
      if (updates.phone) userUpdates.phone = updates.phone;

      if (Object.keys(userUpdates).length > 0) {
        userUpdates.updatedAt = new Date().toISOString();
        await db.collection(USER_COLLECTION_NAME).doc(uid).update(userUpdates);
      }

      return {
        success: true,
        message: "Profile updated successfully",
      };
    } catch (error) {
      logger.error(`❌ [updateProfile] uid=${uid}: ${error.message}`);
      throw error;
    }
  }

  // ─────────────────────────────────────────────
  // PRIVATE HELPERS
  // ─────────────────────────────────────────────

  /**
   * Returns true if the user was created within the last 2 minutes
   * (used to decide whether to send an admin notification).
   */
  async _isFirstTimeUser(uid) {
    try {
      const userDoc = await db.collection(USER_COLLECTION_NAME).doc(uid).get();

      if (!userDoc.exists) return true;

      const data = userDoc.data();
      if (!data?.createdAt) return true;

      const createdAt = data.createdAt.toDate
        ? data.createdAt.toDate()
        : new Date(data.createdAt);

      const diffMinutes = (Date.now() - createdAt.getTime()) / 60_000;
      return diffMinutes < 2;
    } catch {
      return false;
    }
  }

  /**
   * Creates a notification document in adminNotifications so admins can
   * review and approve new handyman registrations.
   * Failures here are logged but never rethrow — registration must not break.
   */
  async _notifyAdminNewHandyman({
    handymanId,
    handymanName,
    email,
    phone,
    profileCompleted,
    timestamp,
  }) {
    try {
      await db.collection(ADMIN_NOTIFICATIONS_COLLECTION_NAME).add({
        type: "new_handyman",
        handymanId,
        handymanName,
        email,
        phone,
        profileCompleted,
        title: profileCompleted
          ? "🎉 New Handyman — Profile Completed"
          : "📝 New Handyman — Profile Incomplete",
        message: profileCompleted
          ? `${handymanName} has registered and completed their profile. Review and approve.`
          : `${handymanName} has registered but skipped profile setup. They may complete it later.`,
        priority: profileCompleted ? "high" : "medium",
        read: false,
        actionRequired: true,
        createdAt: timestamp,
      });

      logger.info(`📧 Admin notification created for handyman: ${handymanId}`);
    } catch (error) {
      // Non-fatal — don't let notification failure break registration
      logger.error(`❌ Failed to create admin notification: ${error.message}`);
    }
  }
}

module.exports = new UserService();

File: lib/services/user_service.dart
import 'package:flutter/foundation.dart';
import 'dart:io';
import 'package:get/get.dart';
import 'package:firebase_auth/firebase_auth.dart';
import 'package:google_sign_in/google_sign_in.dart';
import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:dio/dio.dart';
import 'package:fixilya_app/services/api_client.dart';

class UserService {
  final FirebaseAuth _auth = FirebaseAuth.instance;
  final FirebaseFirestore _firestore = FirebaseFirestore.instance;
  final GoogleSignIn _googleSignIn = GoogleSignIn(scopes: ['email', 'profile']);
  final _api = ApiClient();

  Future<Map<String, dynamic>> completeProfileSkip({
    required String uid,
    required String userType,
  }) async {
    try {
      if (kDebugMode) debugPrint('━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━');
      if (kDebugMode) debugPrint('📝 STARTING complete Profile with Skip WITH BACKEND');
      if (kDebugMode) debugPrint('━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━');
      if (kDebugMode) debugPrint('UID: $uid');
      if (kDebugMode) debugPrint('User Type: $userType');

      final response = await _api.userDio.post(
        '/users/complete-profile-skip',
        data: {'uid': uid, 'userType': userType},
      );

      // response.data will be :
      //  response = {
      //     success: true,
      //     message:
      //       normalizedUserType === "handyman"
      //         ? "Profile created. Complete it later to get approved."
      //         : "Profile created successfully.",
      //     data: {
      //       uid,
      //       userType: normalizedUserType,
      //       profileCompleted: false,
      //       needsApproval: normalizedUserType === "handyman",
      //     },
      //   };

      final data = response.data;
      if (kDebugMode) debugPrint('📥 Backend response: $data');

      if (response.statusCode == 200 && data['success'] == true) {
        if (kDebugMode) debugPrint('✅ Backend Complete Profile Skip successful');

        return {
          'success': true,
          'message':
              data['message'] ?? 'Profile completed with skip successfully',
          'userId': data['data']['uid'],
        };
      } else {
        return {
          'success': false,
          'message': data['message'] ?? 'Profile completion with skip failed',
        };
      }
    } on DioException catch (e) {
      if (kDebugMode) debugPrint('━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━');
      if (kDebugMode) debugPrint('❌ BACKEND API ERROR');
      if (kDebugMode) debugPrint('Status Code: ${e.response?.statusCode}');
      if (kDebugMode) debugPrint('Message: ${e.message}');
      if (kDebugMode) debugPrint('Response: ${e.response?.data}');
      if (kDebugMode) debugPrint('━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━');

      if (e.response != null) {
        final errorData = e.response!.data;
        return {
          'success': false,
          'message':
              errorData['message'] ?? 'Complete Profile With Skip failed',
        };
      } else if (e.type == DioExceptionType.connectionTimeout) {
        return {
          'success': false,
          'message': 'Connection timeout. Please try again.',
        };
      } else if (e.type == DioExceptionType.receiveTimeout) {
        return {
          'success': false,
          'message': 'Server timeout. Please try again.',
        };
      } else {
        return {
          'success': false,
          'message': 'Network error. Please check your connection.',
        };
      }
    } on FirebaseAuthException catch (e) {
      if (kDebugMode) debugPrint('━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━');
      if (kDebugMode) debugPrint('❌ FIREBASE AUTH ERROR');
      if (kDebugMode) debugPrint('Code: ${e.code}');
      if (kDebugMode) debugPrint('Message: ${e.message}');
      if (kDebugMode) debugPrint('━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━');

      return {
        'success': false,
        'message': 'Unknown error',
        'errorCode': e.code,
      };
    } catch (e, stackTrace) {
      if (kDebugMode) debugPrint('━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━');
      if (kDebugMode) debugPrint('❌ UNEXPECTED ERROR');
      if (kDebugMode) debugPrint('Error: $e');
      if (kDebugMode) debugPrint('Stack trace: $stackTrace');
      if (kDebugMode) debugPrint('━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━');

      return {
        'success': false,
        'message': 'An unexpected error occurred. Please try again.',
      };
    }
  }

  Future<Map<String, dynamic>> completeProfile({
    required dynamic user,
    required String userType,
    required String fullName,
    required String phone,
    required String city,
    required String experience,
    required String bio,
    required List<dynamic> skills,
    String? profilePictureUrl,
    List<String> workImageUrls = const [],
  }) async {
    try {
      if (kDebugMode) debugPrint('━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━');
      if (kDebugMode) debugPrint('📝 STARTING complete Profile WITH BACKEND');
      if (kDebugMode) debugPrint('━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━');
      if (kDebugMode) debugPrint('UID: ${user.uid}');
      if (kDebugMode) debugPrint('User Type: $userType');

      final response = await _api.userDio.post(
        '/users/complete-profile',
        data: {
          'uid': user.uid,
          'userType': 'handyman',
          'fullName': user.displayName ?? '',
          'phone': user.phoneNumber ?? '',
          'city': city,
          'experience': experience,
          // 'hourlyRate': double.tryParse(_hourlyRateController.text) ?? 0,
          'bio': bio,
          'skills': skills,
          'profilePicture': profilePictureUrl, // ✅ Cloudinary URL
          'workImages': workImageUrls, // ✅ Cloudinary URLs
        },
      );

      /*
    response will be:
    {
        success: true,
        message: "Your profile is under review. You'll be notified once approved.",
              
        data: {
          uid,
          userType: normalizedUserType,
          profileCompleted: true,
          needsApproval: normalizedUserType === "handyman",
          isNewUser,
          profilePictureUrl: profilePicture || "", // ✅ Return the URL
          workImageUrls: workImages || [], // ✅ Return the URLs
        },
      };
    
  */

      final data = response.data;
      if (kDebugMode) debugPrint('📥 Backend response: $data');

      if (response.statusCode == 200 && data['success'] == true) {
        if (kDebugMode) debugPrint('✅ Backend Complete Profile successful');

        return {
          'success': true,
          'message': data['message'] ?? 'Profile completed successfully',
          'userId': data['data']['uid'],
        };
      } else {
        return {
          'success': false,
          'message': data['message'] ?? 'Profile completion failed',
        };
      }
    } on DioException catch (e) {
      if (kDebugMode) debugPrint('━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━');
      if (kDebugMode) debugPrint('❌ BACKEND API ERROR');
      if (kDebugMode) debugPrint('Status Code: ${e.response?.statusCode}');
      if (kDebugMode) debugPrint('Message: ${e.message}');
      if (kDebugMode) debugPrint('Response: ${e.response?.data}');
      if (kDebugMode) debugPrint('━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━');

      if (e.response != null) {
        final errorData = e.response!.data;
        return {
          'success': false,
          'message': errorData['message'] ?? 'Complete Profile failed',
        };
      } else if (e.type == DioExceptionType.connectionTimeout) {
        return {
          'success': false,
          'message': 'Connection timeout. Please try again.',
        };
      } else if (e.type == DioExceptionType.receiveTimeout) {
        return {
          'success': false,
          'message': 'Server timeout. Please try again.',
        };
      } else {
        return {
          'success': false,
          'message': 'Network error. Please check your connection.',
        };
      }
    } on FirebaseAuthException catch (e) {
      if (kDebugMode) debugPrint('━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━');
      if (kDebugMode) debugPrint('❌ FIREBASE AUTH ERROR');
      if (kDebugMode) debugPrint('Code: ${e.code}');
      if (kDebugMode) debugPrint('Message: ${e.message}');
      if (kDebugMode) debugPrint('━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━');

      return {
        'success': false,
        'message': 'Unknown error',
        'errorCode': e.code,
      };
    } catch (e, stackTrace) {
      if (kDebugMode) debugPrint('━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━');
      if (kDebugMode) debugPrint('❌ UNEXPECTED ERROR');
      if (kDebugMode) debugPrint('Error: $e');
      if (kDebugMode) debugPrint('Stack trace: $stackTrace');
      if (kDebugMode) debugPrint('━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━');

      return {
        'success': false,
        'message': 'An unexpected error occurred. Please try again.',
      };
    }
  }
}
